### PR TITLE
Reorder reduce sum args

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

In https://discourse.mc-stan.org/t/can-we-have-unnormalised-densities-in-user-functions/14333/18 we discussed that the slice argument should come first. This PR does this change.

The release notes section is left empty because this is fixing a "bug" (its more of a prevention from a backwards compatibility nightmare) for a feature that has not been released yet.

## Tests

Changes are reflected in the tests utilities also.

## Side Effects

/

## Checklist

- [x] Math issue #(issue number)

- [x] Copyright holder: Rok Češnovar

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
We could also just fix this on the compiler level for now, though that seems to hacky to me and we have enough time to fix this.

**Comment by wds15** [comment]
This looks all fine to me - I checked everything under stan/math/*/functor/reduce_sum.hpp and here the signatures are all changed as discussed.

ATM the tests are hung up with Linux+MPI for some odd (Jenkins?) reason. Anyway, can we make this PR being tested with a fixed version of the compiler upstream so that it will pass?

I suppose that we have to merge in Stan-math and stanc3 at the same time to get a consistent state.

**Comment by rok-cesnovar** [comment]
> Anyway, can we make this PR being tested with a fixed version of the compiler upstream so that it will pass?

Luckily for this case, the stanc3 is actually not part of the math upstream right now (apart from some but not all parts of cmdstan tests). It does get tested in stanc3 (compiling test models wrt to cmdstan develop).

**Comment by wds15** [comment]
The doc for the reduce_sum_static also needs an update.

File: stan/math/fwd/functor/reduce_sum.hpp [hunk 1]
```diff
@@ -38,7 +38,7 @@ struct reduce_sum_impl {
    * This specialization is not parallelized and works for any autodiff types.
    *
    * An instance, f, of `ReduceFunction` should have the signature:
-   *   T f(int start, int end, Vec&& vmapped_subset, std::ostream* msgs,
+   *   T f(Vec&& vmapped_subset, int start, int end, std::ostream* msgs,
```
Comment (bbbales2) [inline stan/math/fwd/functor/reduce_sum.hpp:41]: Nice you found the Secret Hidden fwd implementation. I am impressed!

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal no need to intervene hear, you said to notify you in case `Cannot contact old-imac: java.lang.InterruptedException` happens. See https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1843/5/pipeline

Restarted already to https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1843/6/pipeline

**Pull request merged**